lark/parser_frontends.py

```diff
@@ -230,31 +187,49 @@ def _prepare_match(self, lexer_conf):
 
             self.regexps[t.name] = lexer_conf.re_module.compile(regexp, lexer_conf.g_regex_flags)
 
-    def parse(self, text, start):
-        return self._parse(start, text)
+    def match(self, term, text, index=0):
+        return self.regexps[term.name].match(text, index)
 
-class XEarley_CompleteLex(XEarley):
-    def __init__(self, *args, **kw):
-        XEarley.__init__(self, *args, complete_lex=True, **kw)
 
+def create_earley_parser__dynamic(lexer_conf, parser_conf, options=None, **kw):
+        earley_matcher = EarleyRegexpMatcher(lexer_conf)
+        return xearley.Parser(parser_conf, earley_matcher.match, ignore=lexer_conf.ignore, **kw)
 
+def _match_earley_basic(term, token):
+    return term.name == token.type
 
-class CYK(WithLexer):
+def create_earley_parser__basic(lexer_conf, parser_conf, options, **kw):
+    return earley.Parser(parser_conf, _match_earley_basic, **kw)
 
-    def __init__(self, lexer_conf, parser_conf, options=None):
-        WithLexer.__init__(self, lexer_conf, parser_conf, options)
-        self.init_traditional_lexer()
+def create_earley_parser(lexer_conf, parser_conf, options):
+    resolve_ambiguity = options.ambiguity == 'resolve'
+    debug = options.debug if options else False
+    tree_class = options.tree_class or Tree if options.ambiguity != 'forest' else None
+
+    extra = {}
+    if lexer_conf.name == 'dynamic':
+        f = create_earley_parser__dynamic
+    elif lexer_conf.name == 'dynamic_complete':
+        extra['complete_lex'] =True
+        f = create_earley_parser__dynamic
+    else:
+        f = create_earley_parser__basic
+
+    return f(lexer_conf, parser_conf, options, resolve_ambiguity=resolve_ambiguity, debug=debug, tree_class=tree_class, **extra)
 
+
+
+class CYK_FrontEnd:
+    def __init__(self, lexer_conf, parser_conf, options=None):
         self._analysis = GrammarAnalyzer(parser_conf)
         self.parser = cyk.Parser(parser_conf.rules)
 
         self.callbacks = parser_conf.callbacks
 
-    def parse(self, text, start):
-        tokens = list(self.make_lexer(text).lex(None))
-        parse = self._parse(start, tokens)
-        parse = self._transform(parse)
-        return parse
+    def parse(self, lexer, start):
```

Comment:
I picked this line as an example: I would suggest clearly making a distinction between variables that contain a Lexer, LexerThread, or LexerState. This function confused me, since `lexer.lex(None)` looks like we are trying to get the Token that get generated from Nothing.

Reply:
I can rename that specific variable to `lexer_thread`.

This doesn't happen anywhere else in the code, only CYK needs special handling.

Other than that, I think everything is named correctly, no?

Reply:
Actually, right now I'm setting `LexerConf.name`, but it could also be a CustomLexer. Any idea for a better name for this attribute?

Reply:
Maybe `lexer_type`?